### DISCUSSION TITLE
Fedimint E-Cash Encoding

### DISCUSSION THREAD

**Discussion by Kodylow** [fedimint/fedimint#3140] [General]
Thoughts on Cashu encoding of ecash tokens for fedimint ecash (hide a fedimint api endpoint in the "mint" endpoint to get the full invite code, try for threshold. allows a receiver to redeem the ecash tokens without really being a part of the federation): 
https://github.com/cashubtc/nuts/blob/main/00.md#token-format

I was thinking we could do something similar to the lightning address flow where the first endpoint you hit gives you a callback url which is where you actually get the invoice. Similarly you could hit the short endpoint we hide in the "mint" encoding, you hit that and it returns json with all the invite codes and data for all the guardians so you can do a proper threshold attempt

It's an extra round vs the standard cashu protocol but we need to get the additional info of the other guardians and I don't think we want to shove all that into the ecash note itself

**Comment by Kodylow** [comment]
@elsirion

**Comment by Kodylow** [comment]
Currently the encoding looks a little like this (collapsed the spendable note parts across files, just trace OOBNotes in the code)

struct OOBNotes {
    pub federation_id: FederationId,
    pub notes: TieredMulti\<SpendableNote\>,
}

struct SpendableNote {
    note: (
        Nonce(secp256k1_zkp::XOnlyPublicKey),
        tbs::Signature
    ),
    spend_key: KeyPair([c_uchar; 96])
}

struct TieredMulti<T>(BTreeMap<Amount, Vec<T>>);


The cashu encoding looks like this:

{
  "token": [
    {
      "mint": "https://8333.space:3338",
      "proofs": [
        {
          "id": "DSAl9nvvyfva",
          "amount": 2,
          "secret": "EhpennC9qB3iFlW8FZ_pZw",
          "C": "02c020067db727d586bc3183aecf97fcb800c3f4cc4759f69c626c9db5d8f5b5d4"
        },
        {
          "id": "DSAl9nvvyfva",
          "amount": 8,
          "secret": "TmS6Cv0YT5PU_5ATVKnukw",
          "C": "02ac910bef28cbe5d7325415d5c263026f15f9b967a079ca9779ab6e5c2db133a7"
        }
      ]
    }
  ],
  "memo": "Thank you."
}


so we'd just coerce into there.

A couple notes: 

"id" is for a specific keyset for the mint because they're trying to implement the rotating audit scheme, so can be omitted or we can hide something there like a federation id maybe?

**Comment by elsirion** [comment]
> [!IMPORTANT]  
> The following assumes this protocol is **only** used for cross-mint offline transfers. For online transfers a direct LN payment flow is always superior as it doesn't make assumptions about recipient wallets.

Our current e-cash format already contains the federation ID since #3024, this means if the client learnt the config of the federation in some way it could validate it and reissued the note already. Now the question is how to discover the federation/download its config:

1. **Include API Url in e-cash note serialization:**
  * Easiest
  * Larger e-cash serialization, especially if we want to run the federation API as a Tor hidden service eventually
  * Some federations might prefer not to be discoverable by outsiders. Once someone has the config there is no way to keep the out due to the privacy guarantees.
2. **Public federation directory:** e.g. via Nostr
  * More complex client implementation/dependencies
  * Another source of unreliability
  * Same e-cash serialization as currently

But even if the federation config can be acquired reliably these payments will be less reliable than online LN or intra-federation offline payments. Even if the sender does not attempt to defraud the recipient due to IRL incentives there might simply not be enough LN liquidity to claim the notes at all times. This means that the recipient has to effectively join the sender's federation till this changes. 

Besides the obvious trust implications this also means running clients for different federations for an indefinite amount of time (so no light client will suffice, it has to be a full one if things fail). This would hugely complicate the current client design or require a wrapper around it that can manage multiple federations.

Given the engineering challenges it would likely be best to showcase the usefulness through a third party implementation before putting too much effort into it from the Fedimint side. Approach 2 mentioned above allows for that without changes to Fedimint.

**Reply by justinmoon** [reply]
> For online transfers a direct LN payment flow is always superior as it doesn't make assumptions about recipient wallets.

The downside of LN payments is that you can't do "push payments" -- recipient always needs to generate invoice before sender can pay. This isn't true for online ecash payments -- the sender can just send the ecash and be done. OFC bolt12 fixes this ...

**Reply by elsirion** [reply]
That's true, BOLT12 without PTLCs will also not be a perfect replacement since there will always be a limit on how many payments can be received on the recipient's behalf before they need to come back online and replenish the encrypted preimage stash.

**Comment by Kodylow** [comment]
Yeah thinking about this a little more doing a "third party receives ecash and immediately cashes out via lightning" would be a big headache for the gateway, just a continuous sink of outgoing Lightning incoming ecash. Cashu's ecash claims are against lightning bitcoin already so the cashout to LN is natural vs our ecash claims are against onchain so involves gateway liquidity like you point out.

I think the more important usecase would be making it cross compatible for cashu ecash clients to hold and store fedimint ecash. 

For the client I was thinking to run less of a client, more of a "just hold the ecash notes" and "establish a 1 time communication with the federation to redeem the ecash notes", not a long lasting relationship. The only thing at rest would be to hold the ecash notes which would contain the API endpoint to hit once you want to redeem, at that time you hit it, get the config, do the redemption.

**Reply by elsirion** [reply]
> Yeah thinking about this a little more doing a "third party receives ecash and immediately cashes out via lightning" would be a big headache for the gateway, just a continuous sink of outgoing Lightning incoming ecash.

From the gateway's perspective the two options

```
sender --e-cash--> GW --LN--> recipient
```

and

```
sender --e-cash--> recipient --e-cash--> GW --LN--> recipient
```

are actually equivalent liquidity-wise. The main drawback is that the recipient now has to be able to process e-cash and can't just be any LN wallet.

> I think the more important usecase would be making it cross compatible for cashu ecash clients to hold and store fedimint ecash.

Why? Can't they just transfer money via LN? E-cash is just an implementation detail from a user's pov, unless there is a clear benefit (e.g. offline transfers) I'd avoid exposing it.

**Reply by dpc** [reply]
Ecash is potentially cheaper, and doesn't rely on some complexity that can fail.

So I wouldn't completely discard the approach. But it might not be a priority RN.

**Reply by elsirion** [reply]
It only moves the complexity to a later point for transactions between different federations. Until the recipient receives the money in *their* federation the transaction cannot be viewed as settled since there is no trust relationship to the sender federation.

For transactions inside the same federation direct e-cash transfers are only marginally less complex than a shortcut LN payment.
